pkg/logs/internal/decoder/auto_multiline_detection/pattern_table.go

```diff
@@ -0,0 +1,133 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+// Package automultilinedetection contains auto multiline detection and aggregation logic.
+package automultilinedetection
+
+import (
+	"github.com/DataDog/datadog-agent/pkg/logs/internal/decoder/auto_multiline_detection/tokens"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+type row struct {
+	tokens    []tokens.Token
+	label     Label
+	count     int64
+	lastIndex int64
+}
+
+// DiagnosticRow is a struct that represents a diagnostic view of a row in the PatternTable.
+type DiagnosticRow struct {
+	TokenString string
+	LabelString string
+	Count       int64
+	LastIndex   int64
+}
+
+// PatternTable is a table of patterns that occur over time from a log source.
+// The pattern table is always sorted by the frequency of the patterns. When the table
+// becomes full, the least recently updated pattern is evicted.
+type PatternTable struct {
+	table          []*row
+	index          int64
```

Comment:
`index`  here just means a monotonically increasing counter right?

Reply:
I struggled to come up with a better name. but mostly yes:
`index` is a monotonic count of each log that passes through the system. 
`lastIndex` is a reference to this same count that represents a point in the past where that pattern last occurred. 

So for the LRU eviction, it acts like a timestamp, such that we can find the least recently seen pattern by looking for the lowest `lastIndex`